Add YAML aware variable substitution

AlphaOne1 opened the pull request:
In the current implementation the environment variable substitution does not follow the specification found in https://opentelemetry.io/docs/specs/otel/configuration/data-model/. Specifically it does not obey the rule that substitution can only occur in YAML values. Further, according to the published specification examples, injected YAML-structured content leads to errors rather than correctly escaped/formatted strings.
This pull request solves these issues by using the yaml.Node structure of the yaml library to operate on already parsed YAML content. Further the redacted examples of the specification are included as further tests.

AlphaOne1 committed:
Add YAML aware environment variable substitution.

Signed-off-by: Alexander Adam <alphaone23@gmail.com>

AlphaOne1 committed:
Improve environment variable substitution handling for robust YAML parsing:

- Adjust `envVarRegexp` to handle multiple dollar signs more effectively.
- Improve node retyping logic for better YAML type detection.
- Remove `checkRawConfType` function and related references.
- Treat injected or invalid YAML types as strings rather than causing errors.
- Update tests to reflect adjusted behavior for YAML injection and substitution handling.

Signed-off-by: Alexander Adam <alphaone23@gmail.com>

pellared commented:
@AlphaOne1, can you please provide examples of current incorrect behavior? It would be nice if you create an issue (bug) for tracking.

pellared commented on otelconf/internal/provider/envprovider.go:
```diff
@@ -20,12 +20,117 @@ const validationPattern = `^[a-zA-Z_][a-zA-Z0-9_]*$`
+	err := yaml.Unmarshal(input, &tree)
+	if err != nil {
+		return nil, err
+	}
```
```suggestion
	if err := yaml.Unmarshal(input, &tree); err != nil {
		return nil, err
	}
```

pellared commented on otelconf/internal/provider/envprovider.go:
```diff
@@ -20,12 +20,117 @@ const validationPattern = `^[a-zA-Z_][a-zA-Z0-9_]*$`
+
+// walkTree recursively traverses the YAML parse tree and replaces environment variables in scalar nodes.
+func walkTree(node *yaml.Node) error {
+	if len(node.Content) == 0 &&
```
is `node.Content) == 0` necessary?

AlphaOne1 committed:
Introduce line-preserving YAML environment variable substitution:

- Add `preserveYAMLLines` to retain original line numbers for improved debugging.
- Replace errors in `retypeNode` with `errors.New` for consistency.
- Enhance test cases to account for escaped newlines in YAML values.

Signed-off-by: Alexander Adam <alphaone23@gmail.com>

# Conflicts:
#	otelconf/config_test.go

pellared reviewed (commented):
I have forgotten to publish some comments 2 weeks ago...

AlphaOne1 committed:
Incorporated comments of @pellared.

Signed-off-by: Alexander Adam <alphaone23@gmail.com>

AlphaOne1 commented on otelconf/internal/provider/envprovider.go:
```diff
@@ -20,12 +20,117 @@ const validationPattern = `^[a-zA-Z_][a-zA-Z0-9_]*$`
+		if child == nil {
+			continue
+		}
+
```
I am not sure about what the patch is exactly. I let a newline normally after blocks for better separation. In other situation with consecutive if, there is also this newline.

codeboten commented on otelconf/internal/provider/envprovider_test.go:
```diff
@@ -224,3 +227,233 @@ func TestReplaceEnvVar(t *testing.T) {
+		},
+		{ // 22
+			yamlInput:  "key: $${STRING_VALUE:-fallback}",
+			yamlOutput: `key: ${STRING_VALUE:-fallback}`, // added missing key:
```
```suggestion
			yamlOutput: `key: ${STRING_VALUE:-fallback}`,
```

codeboten reviewed (commented):
Thanks @AlphaOne1! This looks great, just a few questions in the comments

AlphaOne1 commented on otelconf/internal/provider/envprovider_test.go:
```diff
@@ -224,3 +227,233 @@ func TestReplaceEnvVar(t *testing.T) {
+		},
+		{ // 11
+			yamlInput:  "key: ${INVALID_MAP_VALUE}",
+			yamlOutput: "key: |-\n    value\n    key:value",
```
It is semantically the same, just the YAML library uses a block scalar to serialize the multiline string.

AlphaOne1 committed:
Incorporated comments of @codeboten.

Signed-off-by: Alexander Adam <alphaone23@gmail.com>

AlphaOne1 commented:
I rebased onto main, to resolve the conflicts.

codeboten reviewed (commented):
thanks @AlphaOne1, one of those comments can be resolved
